libff/algebra/curves/bls12_377/bls12_377.sage

```diff
@@ -112,60 +144,79 @@ def g1_fast_subgroup_check_coefficients(g1_order, q, r):
                 # print(f"   p={p}, n={n}")
                 _check_lambdas_not_root_in_subgroup_size(n, c0, c1)
 
-    # Find an element of Fq with multiplicative order 3
+    # Find beta, a primitive cubic root of unity (in Fq\{1}).
+    # Since x^3-1 = (x-1)(x^2+x+1), we find beta by taking the smallest root
+    # of x^2+x+1 mod q returned by `_find_lambda_n` (over the field Fq)
+    # Equivalent to:
+    # [Sage excerpt]
+    # beta = min(root for root in GF(q)(1).nth_root(3, all=True) if root != 1)
     beta = min(_find_lambda_n(GF(q)))
     print(f" beta: {beta}  [in sigma(x,y) = (beta * x, y)]")
+    assert(GF(q)(beta^2) != 1)
+    assert(GF(q)(beta^3) == 1)
 
     # lambda for the r-torsion
-    lambda_r_poly_roots = _find_lambda_n(GF(r));
-    # print(f" lambda_r_poly_roots: {lambda_r_poly_roots}")
-    lambda_r = min(lambda_r_poly_roots) # [root[0] for root in lambda_r_poly_roots])
+    # where r is the scalar field (prime) characteristic,
+    # i.e. the order of the subgroup G1
+    lambda_r = min(_find_lambda_n(GF(r)))
     print(f" (lambda_r: {lambda_r})")
 
+    # TODO: Below replace r by an extra argument `scalar` to represent an arbitrary
+    # scalar in the basis \lambda to use GLV for "generic" scalar mult.
+    # TODO: Implement the GLV scalar decomposition, and benchmark the cost of
+    # computing the endomorphism (GLV truly is of interest when the endomorphism is
+    # efficient as it splits a "big" scalar multiplication into a "small" multi-exponentiation
+    # when Q = sigma(P) is cheap to obtain)
+
     # Find r1_0, r1_1 s.t. r = r1_0 + r1_1*lambda_r
     r1_1 = int(int(r) // int(lambda_r))
     r1_0 = int(r % r1_1)
-    _check_lambdas_not_root_in_subgroups(g1_order, q, r, r1_0, r1_1)
+    _check_lambdas_not_root_in_subgroups(curve_order, q, r, r1_0, r1_1)
 
     return r1_0, r1_1
 
 
 def g1_subgroup_proof_coefficients(curve, r, h1):
     """
     We can prove that some P belongs to G1 by providing some P' in E(Fq) s.t. P
-    = [h1] P'. This requires extra computation upfront, but can be verifed with
+    = [h1] P'. This requires extra computation upfront, but can be verified with
     a single scalar multiplication by h1.
+
+    This is particularly useful to prove subgroup membership in an R1CS program
+    via a multiplication gate.
```

Comment:
> via a scalar multiplication gadget.

I would avoid to refer to "gagdet" here since this truly is an "implementation" term. In arithmetic circuits we talk about gates. I'll use the plural form `via multiplication gates`, thanks

Reply:
> I would avoid to refer to "gagdet" here since this truly is an "implementation" term.
Yes, indeed.  Good point.